In [6]:
import pandas as pd

df1 = pd.read_csv("../data/raw/india_cost_quality_dataset.csv")
df2 = pd.read_csv("../data/raw/livingcost_india_all_inr (4).csv")

print("Dataset 1:", df1.shape)
print("Dataset 2:", df2.shape)

Dataset 1: (170, 7)
Dataset 2: (221, 12)


In [7]:
df1["City"].head(20).tolist()

['Mumbai',
 'Delhi',
 'Bengaluru',
 'Hyderabad',
 'Ahmedabad',
 'Chennai',
 'Kolkata',
 'Pune',
 'Jaipur',
 'Surat',
 'Lucknow',
 'Kanpur',
 'Nagpur',
 'Indore',
 'Thane',
 'Bhopal',
 'Visakhapatnam',
 'Pimpri-Chinchwad',
 'Patna',
 'Vadodara']

In [8]:
df2["City"].head(20).tolist()

[' Adoni',
 ' Anantapur',
 ' Chirala',
 ' Chittoor',
 ' Eluru',
 ' Guntur',
 ' Hindupur',
 ' Kadapa',
 ' Kakinada',
 ' Kurnool',
 ' Machilipatnam',
 ' Madanapalle',
 ' Nandyal',
 ' Nellore',
 ' Ongole',
 ' Proddatur',
 ' Rajahmundry',
 ' Tenali',
 ' Tirupati',
 ' Vijayawada']

In [9]:
df1["City"] = df1["City"].str.strip()
df2["City"] = df2["City"].str.strip()

In [10]:
df2["City"].head(20).tolist()

['Adoni',
 'Anantapur',
 'Chirala',
 'Chittoor',
 'Eluru',
 'Guntur',
 'Hindupur',
 'Kadapa',
 'Kakinada',
 'Kurnool',
 'Machilipatnam',
 'Madanapalle',
 'Nandyal',
 'Nellore',
 'Ongole',
 'Proddatur',
 'Rajahmundry',
 'Tenali',
 'Tirupati',
 'Vijayawada']

In [11]:
common_cities = set(df1["City"]) & set(df2["City"])

len(common_cities)

108

In [12]:
sorted(common_cities)

['Agartala',
 'Agra',
 'Ahmedabad',
 'Aizawl',
 'Ajmer',
 'Aligarh',
 'Allahabad',
 'Alwar',
 'Amritsar',
 'Anantapur',
 'Asansol',
 'Aurangabad',
 'Balurghat',
 'Barasat',
 'Bardhaman',
 'Bareilly',
 'Bhavnagar',
 'Bhilai',
 'Bhopal',
 'Bhubaneswar',
 'Bikaner',
 'Bilaspur',
 'Chandigarh',
 'Chennai',
 'Chittoor',
 'Cuttack',
 'Dehradun',
 'Delhi',
 'Dhanbad',
 'Dibrugarh',
 'Durgapur',
 'Eluru',
 'Faridabad',
 'Gaya',
 'Ghaziabad',
 'Gulbarga',
 'Guntur',
 'Gurgaon',
 'Guwahati',
 'Gwalior',
 'Hisar',
 'Howrah',
 'Imphal',
 'Indore',
 'Jabalpur',
 'Jaipur',
 'Jalandhar',
 'Jalgaon',
 'Jamshedpur',
 'Jhansi',
 'Jodhpur',
 'Kakinada',
 'Kanpur',
 'Kharagpur',
 'Kolhapur',
 'Kolkata',
 'Kollam',
 'Krishnanagar',
 'Kurnool',
 'Lucknow',
 'Ludhiana',
 'Madurai',
 'Malda',
 'Mathura',
 'Meerut',
 'Midnapore',
 'Moradabad',
 'Mumbai',
 'Muzaffarpur',
 'Nagpur',
 'Nanded',
 'Nashik',
 'Navi Mumbai',
 'Nellore',
 'Nizamabad',
 'Pathankot',
 'Patiala',
 'Patna',
 'Pondicherry',
 'Porbandar',
 

In [13]:
cities_df1 = set(df1["City"])
cities_df2 = set(df2["City"])

print("Only in Dataset 1:")
print(sorted(cities_df1 - cities_df2))

print("\nOnly in Dataset 2:")
print(sorted(cities_df2 - cities_df1))

Only in Dataset 1:
['Alipurduar', 'Ambattur', 'Arrah', 'Avadi', 'Bankura', 'Basti', 'Begusarai', 'Belagavi', 'Bengaluru', 'Bettiah', 'Bharatpur', 'Birbhum', 'Bongaigaon', 'Chandrapur', 'Chhapra', 'Cooch Behar', 'Darbhanga', 'Diphu', 'Fatehpur', 'Firozabad', 'Gangarampur', 'Gangtok', 'Haflong', 'Hajipur', 'Haldwani', 'Hapur', 'Hoshiarpur', 'Hubballi-Dharwad', 'Hyderabad', 'Itanagar', 'Jhargram', 'Jorhat', 'Kalyan-Dombivli', 'Katihar', 'Kishanganj', 'Kochi', 'Korba', 'Kota', 'Loni', 'Mandya', 'Mangaluru', 'Medinipur', 'Mira-Bhayandar', 'Morena', 'Motihari', 'Mysuru', 'Nagaon', 'Palakkad', 'Palwal', 'Panipat', 'Pimpri-Chinchwad', 'Purulia', 'Rewa', 'Salem', 'Samastipur', 'Sambalpur', 'Silchar', 'Sivasagar', 'Siwan', 'Tezpur', 'Tinsukia', 'Vasai-Virar']

Only in Dataset 2:
['Adoni', 'Akola', 'Alleppey', 'Amravati', 'Bally', 'Bangalore', 'Baramulla', 'Baranagar', 'Barrackpore', 'Belgaum', 'Berhampore', 'Bhagalpur', 'Bhatpara', 'Bhilwara', 'Bhiwandi', 'Bhiwani', 'Bhuj', 'Bidhannagar', 'Bijap

1. Clear naming mismatches

These should be normalized because they are obviously the same city:

Dataset 1	Dataset 2
Bengaluru	Bangalore
Belagavi	Belgaum
Mangaluru	Mangalore
Mysuru	Mysore

In [14]:
city_mapping = {
    "Bangalore": "Bengaluru",
    "Belgaum": "Belagavi",
    "Mangalore": "Mangaluru",
    "Mysore": "Mysuru",
    "Kochi, Kerala": "Kochi",
    "Hyderabad, Telangana": "Hyderabad",
    "Kota, Rajasthan": "Kota",
    "Salem, Tamil Nadu": "Salem"
}

df1["City"] = df1["City"].replace(city_mapping)
df2["City"] = df2["City"].replace(city_mapping)

replaced the contradicting spelling or name of the same cities in both datasets

In [15]:
common_cities = set(df1["City"]) & set(df2["City"])

print("Common cities:", len(common_cities))

Common cities: 116


In [16]:
print(sorted(common_cities))

['Agartala', 'Agra', 'Ahmedabad', 'Aizawl', 'Ajmer', 'Aligarh', 'Allahabad', 'Alwar', 'Amritsar', 'Anantapur', 'Asansol', 'Aurangabad', 'Balurghat', 'Barasat', 'Bardhaman', 'Bareilly', 'Belagavi', 'Bengaluru', 'Bhavnagar', 'Bhilai', 'Bhopal', 'Bhubaneswar', 'Bikaner', 'Bilaspur', 'Chandigarh', 'Chennai', 'Chittoor', 'Cuttack', 'Dehradun', 'Delhi', 'Dhanbad', 'Dibrugarh', 'Durgapur', 'Eluru', 'Faridabad', 'Gaya', 'Ghaziabad', 'Gulbarga', 'Guntur', 'Gurgaon', 'Guwahati', 'Gwalior', 'Hisar', 'Howrah', 'Hyderabad', 'Imphal', 'Indore', 'Jabalpur', 'Jaipur', 'Jalandhar', 'Jalgaon', 'Jamshedpur', 'Jhansi', 'Jodhpur', 'Kakinada', 'Kanpur', 'Kharagpur', 'Kochi', 'Kolhapur', 'Kolkata', 'Kollam', 'Kota', 'Krishnanagar', 'Kurnool', 'Lucknow', 'Ludhiana', 'Madurai', 'Malda', 'Mangaluru', 'Mathura', 'Meerut', 'Midnapore', 'Moradabad', 'Mumbai', 'Muzaffarpur', 'Mysuru', 'Nagpur', 'Nanded', 'Nashik', 'Navi Mumbai', 'Nellore', 'Nizamabad', 'Pathankot', 'Patiala', 'Patna', 'Pondicherry', 'Porbandar', 'P

In [17]:
print("Only in Dataset 1:")
print(sorted(set(df1["City"]) - set(df2["City"])))

print("\nOnly in Dataset 2:")
print(sorted(set(df2["City"]) - set(df1["City"])))

Only in Dataset 1:
['Alipurduar', 'Ambattur', 'Arrah', 'Avadi', 'Bankura', 'Basti', 'Begusarai', 'Bettiah', 'Bharatpur', 'Birbhum', 'Bongaigaon', 'Chandrapur', 'Chhapra', 'Cooch Behar', 'Darbhanga', 'Diphu', 'Fatehpur', 'Firozabad', 'Gangarampur', 'Gangtok', 'Haflong', 'Hajipur', 'Haldwani', 'Hapur', 'Hoshiarpur', 'Hubballi-Dharwad', 'Itanagar', 'Jhargram', 'Jorhat', 'Kalyan-Dombivli', 'Katihar', 'Kishanganj', 'Korba', 'Loni', 'Mandya', 'Medinipur', 'Mira-Bhayandar', 'Morena', 'Motihari', 'Nagaon', 'Palakkad', 'Palwal', 'Panipat', 'Pimpri-Chinchwad', 'Purulia', 'Rewa', 'Samastipur', 'Sambalpur', 'Silchar', 'Sivasagar', 'Siwan', 'Tezpur', 'Tinsukia', 'Vasai-Virar']

Only in Dataset 2:
['Adoni', 'Akola', 'Alleppey', 'Amravati', 'Bally', 'Baramulla', 'Baranagar', 'Barrackpore', 'Berhampore', 'Bhagalpur', 'Bhatpara', 'Bhilwara', 'Bhiwandi', 'Bhiwani', 'Bhuj', 'Bidhannagar', 'Bijapur', 'Chakradharpur', 'Chandannagar', 'Chirala', 'Coimbatore', 'Cuddalore', 'Davanagere', 'Dindigul', 'Dispur',

For the same city, how different are Dataset 1 and Dataset 2?

In [18]:
common_cities = sorted(set(df1["City"]) & set(df2["City"]))

df1_common = df1[df1["City"].isin(common_cities)].copy()
df2_common = df2[df2["City"].isin(common_cities)].copy()

print("Dataset 1:", df1_common.shape)
print("Dataset 2:", df2_common.shape)

Dataset 1: (116, 7)
Dataset 2: (116, 12)


In [19]:
print("Dataset 1 columns:")
print(df1_common.columns.tolist())

print("\nDataset 2 columns:")
print(df2_common.columns.tolist())

Dataset 1 columns:
['City', 'Average Rent (INR/month)', 'Food Cost (INR/month)', 'Internet Speed (Mbps)', 'Healthcare Rating', 'Safety Score', 'Happiness Index']

Dataset 2 columns:
['City', 'cost_one_person_usd', 'rent_one_person_usd', 'monthly_salary_after_tax_usd', 'income_after_rent_usd', 'months_covered', 'cost_one_person_inr', 'rent_one_person_inr', 'monthly_salary_after_tax_inr', 'income_after_rent_inr', 'usd_to_inr_rate_used', 'source_url']


In [20]:
df1_common[
    ["Average Rent (INR/month)", "Food Cost (INR/month)",
     "Internet Speed (Mbps)", "Healthcare Rating",
     "Safety Score", "Happiness Index"]
].describe()

,Average Rent (INR/month),Food Cost (INR/month),Internet Speed (Mbps),Healthcare Rating,Safety Score,Happiness Index
count,116.000000,116.000000,116.000000,116.000000,116.000000,116.000000
mean,8483.543103,5503.163793,85.833190,6.424138,5.323276,6.338793
std,5685.971878,1416.506799,33.809591,1.433428,1.992916,1.311605
min,4009.000000,3021.000000,20.350000,4.000000,2.100000,4.100000
25%,5498.500000,4397.500000,58.640000,5.175000,3.675000,5.175000
50%,6876.000000,5439.000000,90.105000,6.350000,5.400000,6.500000
75%,8580.000000,6716.000000,110.657500,7.700000,6.950000,7.500000
max,34896.000000,7999.000000,149.460000,8.800000,8.900000,8.400000


In [21]:
df2_common[
    ["cost_one_person_inr",
     "rent_one_person_inr",
     "monthly_salary_after_tax_inr",
     "income_after_rent_inr",
     "months_covered"]
].describe()

,cost_one_person_inr,rent_one_person_inr,monthly_salary_after_tax_inr,income_after_rent_inr,months_covered
count,116.000000,116.000000,116.000000,116.000000,116.000000
mean,31643.086466,10866.252069,32318.729397,21452.477500,1.100862
std,7944.760601,5173.845961,9045.532879,10398.731685,0.357891
min,22053.810000,5298.190000,17484.900000,-16430.530000,0.300000
25%,26315.207500,8061.507500,25722.125000,15778.142500,0.900000
50%,29346.510000,9708.950000,31499.175000,21803.400000,1.100000
75%,33717.737500,11378.362500,37320.147500,27336.627500,1.300000
max,61943.980000,40241.620000,59747.380000,44371.220000,2.100000


In [22]:
merged_df = pd.merge(
    df1_common,
    df2_common,
    on="City",
    how="inner"
)

merged_df.shape

(116, 18)

In [23]:
merged_df.columns.tolist()

['City',
 'Average Rent (INR/month)',
 'Food Cost (INR/month)',
 'Internet Speed (Mbps)',
 'Healthcare Rating',
 'Safety Score',
 'Happiness Index',
 'cost_one_person_usd',
 'rent_one_person_usd',
 'monthly_salary_after_tax_usd',
 'income_after_rent_usd',
 'months_covered',
 'cost_one_person_inr',
 'rent_one_person_inr',
 'monthly_salary_after_tax_inr',
 'income_after_rent_inr',
 'usd_to_inr_rate_used',
 'source_url']

Now we clean the merged dataset

For our India-focused analysis, we don't need the USD duplicates, exchange-rate column, or source URL as model features.

In [24]:
model_df = merged_df.drop(columns=[
    "cost_one_person_usd",
    "rent_one_person_usd",
    "monthly_salary_after_tax_usd",
    "income_after_rent_usd",
    "usd_to_inr_rate_used",
    "source_url"
])

model_df.shape

(116, 12)

In [25]:
model_df.columns.tolist()

['City',
 'Average Rent (INR/month)',
 'Food Cost (INR/month)',
 'Internet Speed (Mbps)',
 'Healthcare Rating',
 'Safety Score',
 'Happiness Index',
 'months_covered',
 'cost_one_person_inr',
 'rent_one_person_inr',
 'monthly_salary_after_tax_inr',
 'income_after_rent_inr']

In [26]:
model_df.isnull().sum()

City                            0
Average Rent (INR/month)        0
Food Cost (INR/month)           0
Internet Speed (Mbps)           0
Healthcare Rating               0
Safety Score                    0
Happiness Index                 0
months_covered                  0
cost_one_person_inr             0
rent_one_person_inr             0
monthly_salary_after_tax_inr    0
income_after_rent_inr           0
dtype: int64

In [27]:
model_df["City"].duplicated().sum()

np.int64(0)

In [28]:
model_df[[
    "City",
    "Average Rent (INR/month)",
    "rent_one_person_inr"
]].head(20)

,City,Average Rent (INR/month),rent_one_person_inr
0,Mumbai,34896,35233.38
1,Delhi,26424,16694.12
2,Bengaluru,26667,19417.90
3,Hyderabad,25785,14936.85
4,Ahmedabad,20881,13970.34
5,Chennai,21323,13970.34
6,Kolkata,25710,10631.52
7,Pune,33540,18363.53
8,Jaipur,8774,10982.98
9,Surat,9697,40241.62


In [29]:
Don't delete either column yet.

Instead, let's quantify their relationship across all 116 cities.

SyntaxError: unterminated string literal (detected at line 1) (1239029306.py, line 1)

In [30]:
model_df[
    ["Average Rent (INR/month)", "rent_one_person_inr"]
].corr()

,Average Rent (INR/month),rent_one_person_inr
Average Rent (INR/month),1.000000,0.417648
rent_one_person_inr,0.417648,1.000000


In [31]:
model_df["rent_difference_inr"] = (
    model_df["rent_one_person_inr"]
    - model_df["Average Rent (INR/month)"]
)

model_df["rent_difference_percent"] = (
    abs(model_df["rent_difference_inr"])
    / model_df["Average Rent (INR/month)"]
) * 100

In [32]:
model_df[
    ["City",
     "Average Rent (INR/month)",
     "rent_one_person_inr",
     "rent_difference_percent"]
].sort_values(
    "rent_difference_percent",
    ascending=False
).head(15)

,City,Average Rent (INR/month),rent_one_person_inr,rent_difference_percent
60,Belagavi,4557,21965.95,382.026553
63,Nellore,6394,26534.87,314.996403
9,Surat,9697,40241.62,314.990409
74,Dehradun,4151,14058.21,238.670441
14,Thane,6315,20472.27,224.184798
50,Gurgaon,5961,18890.72,216.905217
31,Navi Mumbai,7579,22932.45,202.578836
24,Meerut,4065,11861.61,191.798524
102,Dibrugarh,4299,12125.20,182.046988
61,Mangaluru,5254,14146.07,169.243814


In [ ]:
model_df["rent_difference_percent"].describe()

count    116.000000
mean      62.529602
std       69.778630
min        0.275110
25%       17.468612
50%       37.738693
75%       84.882910
max      382.026553
Name: rent_difference_percent, dtype: float64

In [ ]:
model_df[
    ["City",
     "Average Rent (INR/month)",
     "rent_one_person_inr",
     "rent_difference_percent"]
].sort_values(
    "rent_difference_percent",
    ascending=False
).head(15)

,City,Average Rent (INR/month),rent_one_person_inr,rent_difference_percent
60,Belagavi,4557,21965.95,382.026553
63,Nellore,6394,26534.87,314.996403
9,Surat,9697,40241.62,314.990409
74,Dehradun,4151,14058.21,238.670441
14,Thane,6315,20472.27,224.184798
50,Gurgaon,5961,18890.72,216.905217
31,Navi Mumbai,7579,22932.45,202.578836
24,Meerut,4065,11861.61,191.798524
102,Dibrugarh,4299,12125.20,182.046988
61,Mangaluru,5254,14146.07,169.243814


⚠️ One important observation

Don't treat rent_difference_percent as a direct "rent increase" or actual cost difference. The two datasets likely use different definitions/sources for rent, which explains extreme values like 382%.

So I'd keep it, but treat it as a derived comparison feature, not as ground truth.

Next step

Before we start modeling, let's finish the EDA/feature understanding properly.

In [ ]:
model_df.to_csv("../data/processed/india_cost_quality_merged.csv", index=False)

In [ ]:
model_df.shape

(116, 14)

03_data_Integration.ipynb has covered the complete integration stage:

Loaded both raw datasets
Inspected their city names
Identified naming inconsistencies
Compared common/unique cities
Established the 116 reliable common cities
Merged the datasets
Removed USD/source columns that aren't needed for the model
Checked missing values → 0
Compared the two rent measures
Created rent_difference_inr and rent_difference_percent
Saved the final processed dataset:
data/processed/india_cost_quality_merged.csv
Final dataset → 116 × 14